drivers/sensor/max17055/max17055.c

```diff
@@ -0,0 +1,245 @@
+
+	return rem;
+}
+
+static void set_millis(struct sensor_value *val, int val_millis)
+{
+	val->val1 = val_millis / 1000;
+	val->val2 = val_millis % 1000 * 1000;
+}
+
+/**
+ * @brief sensor value get
+ *
+ * @return -ENOTSUP for unsupported channels
+ */
+static int max17055_channel_get(struct device *dev, enum sensor_channel chan,
+				struct sensor_value *val)
+{
+	const struct max17055_config *const config = dev->config_info;
+	struct max17055_data *max17055 = dev->driver_data;
+	unsigned int tmp;
+
+	switch (chan) {
+	case SENSOR_CHAN_GAUGE_VOLTAGE:
+		/* Get voltage in uV */
+		tmp = max17055->voltage * 1250 / 16;
+		val->val1 = tmp / 1000000;
+		val->val2 = tmp % 1000000;
+		break;
+	case SENSOR_CHAN_GAUGE_AVG_CURRENT: {
+		int cap_ma;
+
+		cap_ma = capacity_to_ma(config->rsense_mohms,
+					max17055->avg_current);
+		set_millis(val, cap_ma);
+		break;
+	}
+	case SENSOR_CHAN_GAUGE_STATE_OF_CHARGE:
+		val->val1 = max17055->state_of_charge / 256;
+		val->val2 = max17055->state_of_charge % 256;
+		break;
+	case SENSOR_CHAN_GAUGE_TEMP:
+		val->val1 = max17055->internal_temperature / 256;
+		val->val2 = max17055->internal_temperature % 256 *
+			(1000000 / 256);
+		break;
+	case SENSOR_CHAN_GAUGE_FULL_CHARGE_CAPACITY:
+		tmp = capacity_to_ma(config->rsense_mohms,
+				     max17055->full_charge_capacity);
+		set_millis(val, tmp);
+		break;
+	case SENSOR_CHAN_GAUGE_REMAINING_CHARGE_CAPACITY:
+		tmp = capacity_to_ma(config->rsense_mohms,
+				     max17055->remaining_charge_capacity);
+		set_millis(val, tmp);
+		break;
+	case SENSOR_CHAN_GAUGE_TIME_TO_EMPTY:
+		/* Get time in ms */
+		if (max17055->time_to_empty == 0xffff) {
+			val->val1 = 0;
+			val->val2 = 0;
+		} else {
+			tmp = max17055->time_to_empty * 5625;
+			set_millis(val, tmp);
+		}
+		break;
+	case SENSOR_CHAN_GAUGE_TIME_TO_FULL:
+		if (max17055->time_to_full == 0xffff) {
+			val->val1 = 0;
+			val->val2 = 0;
+		} else {
+			/* Get time in ms */
+			tmp = max17055->time_to_full * 5625;
+			set_millis(val, tmp);
+		}
+		break;
+	case SENSOR_CHAN_GAUGE_CYCLE_COUNT:
+		val->val1 = max17055->cycle_count / 100;
+		val->val2 = max17055->cycle_count % 100 * 10000;
+		break;
+	case SENSOR_CHAN_GAUGE_NOM_AVAIL_CAPACITY:
+		tmp = capacity_to_ma(config->rsense_mohms,
+				     max17055->design_capacity);
+		set_millis(val, tmp);
+		break;
+	case SENSOR_CHAN_GAUGE_DESIGN_VOLTAGE:
+		set_millis(val, config->design_voltage);
+		break;
+	case SENSOR_CHAN_GAUGE_DESIRED_VOLTAGE:
+		set_millis(val, config->desired_voltage);
+		break;
+	case SENSOR_CHAN_GAUGE_DESIRED_CHARGING_CURRENT:
+		val->val1 = config->desired_charging_current;
+		val->val2 = 0;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int max17055_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	struct max17055_data *max17055 = dev->driver_data;
+	int status = 0;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+
+	status = max17055_reg_read(max17055, VCELL, &max17055->voltage);
+	if (status) {
```

Comment:
Should be `if (status < 0)` given that the `max17055_reg_read` function returns `-Exxx`.

Reply:
I don't agree with that, sorry. It cannot return a positive value, so any non-zero value is negative and an error

Reply:
If it is likely the function will never return any meaningful positive value (e.g. length of read data) even in the future, then please change this to `if (status != 0)`.

Reply:
It is currently:

if (status)

which is equivalent and less verbose. It looks like both approaches are used throughout Zephyr although the one you suggest is a little less common. Is this a style thiing?

zephyr$ git grep "if (ret)" |wc
    530    2131   24728
zephyr$ git grep "if (result)" |wc
     10      40     473
zephyr$ git grep "if (rc)" |wc
    139     556    5699
zephyr$ git grep "if (rc != 0)" |wc
    182    1121    9870
zephyr$ git grep "if (ret != 0)" |wc
    166     997    8738
zephyr$ git grep "if (result != 0)" |wc
      5      34     264

Reply:
Given that `status` is not semantically boolean, it is desirable to properly specify the expression such that the result is boolean and there is no room for any ambiguity.

In general, implicit integer to boolean conversion/interpretation is considered a bad practice.

Reply:
I'm not sure about that. C has only got a bool type relatively recently. if() has always been used to check for non-zero.

But I've changed it since you seem pretty sure of it.

Reply:
> But I've changed it since you seem pretty sure of it.

Thanks for doing that. Such constructs are already being enforced by various maintainers, although there is no official guideline/rule about it at this time (we really should have one in the long run).

MISRA-C and other functional safety standard compliance will eventually enforce this too, although the specifics may be a little different (e.g. MISRA C:2012 Rule 14.4).